### PR TITLE
vendor: migrate to "golang/dep"

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/7225.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -20,3 +20,20 @@ hack/scripts-dev/docker-dns/.Dockerfile
...
+!vendor/**/*.cpp
+!vendor/**/*.s
+!vendor/**/COPYING*
+!vendor/**/PATENTS*
+!vendor/**/NOTICE*
+!vendor/**/Licence*
+!vendor/**/License*
+!vendor/**/LICENCE*
+!vendor/**/LICENSE*
+vendor/**/*_test.go
```
Comment (xiang90) [inline .gitignore]: new line

Comment (gyuho) [reply .gitignore]: Just added.

**Comment by gyuho** [comment]
Major change would be enforcing etcd dependencies (e.g. grpc-go v1.7.5 on `clientv3`) on `go get` commands. Previously, we didn't do it to let users choose their own gRPC version. Now, clientv3 is so sensitive to gRPC versions, so it has to be enforced. Also, Go 1.9 type alias won't complain about context import conflicts.

@olivere @universonic Could you try this out? Thanks!

**Comment by gyuho** [comment]
@sdboyer Could you quickly go over this?

Gopkg.toml https://github.com/coreos/etcd/pull/9155/commits/35cae902c7dbf646cada5fe97eab1660047d18ba
scripts/updatedep.sh https://github.com/coreos/etcd/pull/9155/commits/9a75a2a664f8580edcadde8e87041e2fc5aae4c8

https://github.com/coreos/etcd/pull/9155/commits/1838ee19be8a10c28294aaafbd689c29e07e6ada is the diff between `glide` + `glide vc --only-code --no-tests` and `dep` with `gitignore`.

`dep` is still complete prune feature (e.g. `github.com/golang/groupcache/...` is directive dependency, but not `github.com/golang/groupcache` itself). However, I think we are ok with it. Other than that, `golang/dep` works great!

File: build [hunk 2]
```diff
@@ -36,31 +36,14 @@ etcd_build() {
 	# Static compilation is useful when etcd is run in a container. $GO_BUILD_FLAGS is OK
 
 	# shellcheck disable=SC2086
-	CGO_ENABLED=0 go build $GO_BUILD_FLAGS -installsuffix cgo -ldflags "$GO_LDFLAGS" -o "${out}/etcd" ${REPO_PATH}/cmd/etcd || return
+	CGO_ENABLED=0 go build $GO_BUILD_FLAGS -installsuffix cgo -ldflags "$GO_LDFLAGS" -o "${out}/etcd" . || return
```
Comment (fanminshi) [inline build]: i think having `${REPO_PATH}` as build path is more strict.

File: build [hunk 3]
```diff
@@ -36,31 +36,14 @@ etcd_build() {
...
-	etcdGOPATH="${CDIR}/gopath"
-	# preserve old gopath to support building with unvendored tooling deps (e.g., gofail)
-	if [ -n "$GOPATH" ]; then
-		GOPATH=":$GOPATH"
-	fi
-	export GOPATH=${etcdGOPATH}$GOPATH
-	rm -rf "${etcdGOPATH}/src"
-	mkdir -p "${etcdGOPATH}"
-	ln -s "${CDIR}/cmd/vendor" "${etcdGOPATH}/src"
+	CGO_ENABLED=0 go build $GO_BUILD_FLAGS -installsuffix cgo -ldflags "$GO_LDFLAGS" -o "${out}/etcdctl" ./etcdctl || return
```
Comment (fanminshi) [inline build]: maybe have `${REPO_PATH}/etcdctl`? i think it is more strict.

**Comment by fanminshi** [comment]
lgtm after few suggestions.

**Comment by gyuho** [comment]
All fixed. Will merge on greens.

**Comment by sdboyer** [comment]
ack. sorry. i was taken up all this week with [the new release of dep](https://golang.github.io/dep/docs/introduction.html). awesome to see you pulled the trigger, though!

the comments i'd make are:

1. it's redundant, and potentially harmful, to specify `source` in `Gopkg.toml` when it's not actually needed (and it's not in any of the cases i saw from skimming).
2. unless there's a specific reason that you need _exactly_ those revisions for all those projects, and that they NEVER change under ANY circumstances, it's preferable that you instead just specify e.g. `branch = "master"`, and trust `Gopkg.lock` to give you the version stability you want.
3. v0.4.1 of dep prunes automatically, and you have [more granular controls](https://golang.github.io/dep/docs/Gopkg.toml.html#prune) over it from within `Gopkg.toml`.
4. overrides should also be avoided, when possible. they can be useful as a one-time way of getting `Gopkg.lock` to the right place, but once you've done that, you should ideally be able to remove them.
5. strongly prefer released versions to `go get`-ing tip. we should have a curl-able installer soon https://github.com/golang/dep/pull/1533

**Comment by gyuho** [comment]
@sdboyer Awesome. Thanks for the tips! We will address those shortly!

**Comment by tamird** [comment]
There are some really unfortunate side effects as a result of this. For instance, cockroachdb wants to use a newer version of gogo/protobuf than what's specified in your new Gopkg.toml - the result is that dep will refuse to ignore etcd's specified version and won't resolve (in the presence of the two constraints etcd.branch = master and protobuf.branch = master).

As far as I know, there's no workaround here, other than for etcd to be very careful to not over-constrain your dependencies (and currently they appear to all be maximally constrained; your Gopkg.toml is roughly equivalent to your Gopkg.lock).

**Comment by xiang90** [comment]
@tamird 

can you move the discussion to here https://github.com/coreos/etcd/issues/9224? 

thanks!

**Comment by gyuho** [comment]
@tamird Wasn't aware of it. We *did* over-constrain the dependencies to minimize the difference during migration. But open to any workaround if that works for external projects.

I see https://github.com/cockroachdb/cockroach/blob/master/Gopkg.toml#L92

```toml
[[constraint]]
  name = "github.com/gogo/protobuf"
  branch = "master"
```

Would `[[override]]` work for `cockroachdb/cockroach`?

**Comment by tamird** [comment]
@gyuho no, that still doesn't work.

You can, at minimum, remove all `[constraint]`s from Gopkg.toml now that you have a Gopkg.lock.

**Comment by gyuho** [comment]
Replacing all our constraints with `branch = "master"` would work?

**Comment by sdboyer** [comment]
why does an override not work? this is what they're actually designed for.

**Comment by tamird** [comment]
@gyuho you shouldn't need any constraints at all, provided you don't run `dep ensure -update`. The Gopkg.lock will prevent any of the versions from changing.

I'm suggesting you simply delete all the `[constraint]`s without replacement.

@sdboyer "why" is a bit of an academic question - I can't tell you why. All I know is that using `[override]` produces the exact same result as using `[constraint]` in this case. Here's the key line of output, in case it's helpful:

> Solving failure: No versions of github.com/coreos/etcd met constraints: 
...
        master: Could not introduce github.com/coreos/etcd@master, as it depends on github.com/gogo/protobuf from https://github.com/gogo/protobuf, but github.com/gogo/protobuf is already marked as coming from github.com/gogo/protobuf by github.com/cockroachdb/cockroach

**Comment by sdboyer** [comment]
I'm not sure what makes "why" an academic question - that's the info i needed. 

the problem is a conflict on `source`. you can also issue an override on that.

**Comment by gyuho** [comment]
I'm fine with deleting constraints, but that means etcd will need hand-maintain `Gopkg.lock`, which is also fine with me:

> Gopkg.lock is autogenerated; editing it manually is generally an antipattern. If there is a goal you can only achieve by hand-editing Gopkg.lock, it is at least a feature request, and likely a bug.

@sdboyer Is this the valid use case for editing `Gopkg.lock`? Or I'm missing something.

**Comment by tamird** [comment]
@sdboyer yeah, that seems to work, but requires an override for every overlapping dependency (of which github.com/gogo/protobuf is just one).

@gyuho why do you need to hand-edit Gopkg.lock? You can update single packages with `dep ensure -update <pkg>`. Are you often updating to a version that is not the latest?

**Pull request merged**